### PR TITLE
Clear queue when an error occurs during sync

### PR DESCRIPTION
**Pull request by simonrozsival**
Ref #661 

This PR makes sure we don't immediately continue with synchronization when an error is reported. Please review the thread-safety test twice, just for sure.. 😃 

Squash whenever

### CODE REVIEW

File: Toggl.Foundation.Tests/Sync/SyncStateQueueTests.cs [hunk 1]
```diff
@@ -31,119 +33,166 @@ private void queueSyncs(params SyncState[] states)
...
+
+                Clear();
+                var returnValues = DequeueUntilSleep();
+
+                returnValues.Should().HaveCount(1);
+                returnValues.ShouldBeSameEventsAs(Sleep);
+            }
+
+            [Property]
+            public void RetursSleepAfterClearNoMatterWhatWasQueuedPreviously(bool[] pushPull)
```
Comment (willsb) [inline Toggl.Foundation.Tests/Sync/SyncStateQueueTests.cs]: `NonEmptyArray<bool>`, maybe?

**Comment by amulware** [comment]
Given that neither of us reviewed the tests well, I'll go over those again in a minute.

**Comment by simonrozsival** [comment]
@amulware That can happen and I don't think it would be incorrect behavior. We might temporarily block further syncing, but what should be the reason to enable it again? A timeout? Force sync?

**Comment by amulware** [comment]
Well, the thing is that the UI causes the sync not knowing that it shouldn't anymore. It's up to the UI to figure out if it thinks syncing works again, but because of this race condition, I don't see that working. Though I'm not sure if this can actually be avoided at all.

**Comment by amulware** [comment]
Unless we can somehow enforce that no sync is scheduled until the error is handled on the ui thread or something. Or maybe we need to temporarily freeze the state machine, and the unfreeze it manually in the ui.

But I'm not sure it's worth all these complications. Let's maybe leave it how it is and see if we notice anything bad happening because of this. (Maybe graphical glitches, or weird double navigation to the error screens?)

**Comment by simonrozsival** [comment]
Then maybe besides clearing the queue, we can block further syncing for the next `n` seconds, or until user forces sync manually.

The error might be being offline (probably most frequent one), timeout (server is not available for some reason and retry loop reaches a delay > 60s), or if there is a bug in the app. 410, 418 and 401 will destroy (freeze) the `SyncManager`, so that is not a case which we need to consider.

I think we should leave this as is right now, and maybe create an issue to consider possible solutions for this problem later.

**Comment by amulware** [comment]
In the case of offline and timeout, I dont really care, only 401, 410 and 418 are problematic to me. But if you say that you will freeze in that case, then maybe it's mostly alright? Or is the UI going to decide on that freeze? If so there is still a race condition... But let's leave this as is for now and we'll see.

**Comment by simonrozsival** [comment]
If we run into these 3 cases, the user has to either update the app and start it again or log in again (=> new data source with new sync manager). Therefore we should freeze it in those cases - it is already prepared in #978.

File: Toggl.Foundation.Tests/Sync/SyncStateQueueTests.cs [hunk 2]
```diff
@@ -31,119 +34,166 @@ private void queueSyncs(params SyncState[] states)
...
+            [Property]
+            public void RetursSleepAfterClearNoMatterWhatWasQueuedPreviously(NonEmptyArray<bool> pushPull)
+            {
+                DequeuedStates.Clear();
+                QueueSyncs(BoolToEnumValues(pushPull.Get));
+
+                Clear();
+                var returnValues = DequeueUntilSleep();
+
+                returnValues.Should().HaveCount(1);
```
Comment (amulware) [inline Toggl.Foundation.Tests/Sync/SyncStateQueueTests.cs]: Are these count-checks necessary?

**Pull request merged**